azure-servicebus/src/main/java/com/microsoft/azure/servicebus/IMessageAndSessionPump.java

```diff
@@ -16,41 +19,96 @@
+
+    /**
+     * Receive session messages continuously from the queue. Registers a message handler and begins a new thread to receive session-messages.
+     * ISessionHandler methods are executed on java.util.concurrent.commonPool()
+     * 
      * @param handler The {@link ISessionHandler} instance
      * @throws InterruptedException if the current thread was interrupted while waiting
      * @throws ServiceBusException  if register failed
      */
+    @Deprecated
     public void registerSessionHandler(ISessionHandler handler) throws InterruptedException, ServiceBusException;
+    
+    /**
+     * Receive session messages continuously from the queue. Registers a message handler and begins a new thread to receive session-messages.
+     * ISessionHandler methods are executed on the passed executor service.
+     *
+     * @param handler The {@link ISessionHandler} instance
+     * @param executorService ExecutorService which is used to execute {@link ISessionHandler} methods
+     * @throws InterruptedException if the current thread was interrupted while waiting
+     * @throws ServiceBusException  if register failed
+     */
+    public void registerSessionHandler(ISessionHandler handler, ExecutorService executorService) throws InterruptedException, ServiceBusException;
 
     /**
-     * Receive session messages continously from the queue. Registers a message handler and begins a new thread to receive session-messages.
+     * Receive session messages continuously from the queue. Registers a message handler and begins a new thread to receive session-messages.
+     * ISessionHandler methods are executed on java.util.concurrent.commonPool()
      *
      * @param handler        The {@link ISessionHandler} instance
      * @param handlerOptions {@link SessionHandlerOptions}
      * @throws InterruptedException if the current thread was interrupted while waiting
      * @throws ServiceBusException  if register failed
      */
+    @Deprecated
     public void registerSessionHandler(ISessionHandler handler, SessionHandlerOptions handlerOptions) throws InterruptedException, ServiceBusException;
+    
+    /**
+     * Receive session messages continuously from the queue. Registers a message handler and begins a new thread to receive session-messages.
+     * ISessionHandler methods are executed on the passed executor service.
+     *
+     * @param handler        The {@link ISessionHandler} instance
+     * @param handlerOptions {@link SessionHandlerOptions}
+     * @param executorService ExecutorService which is used to execute {@link ISessionHandler} methods
+     * @throws InterruptedException if the current thread was interrupted while waiting
+     * @throws ServiceBusException  if register failed
+     */
+    public void registerSessionHandler(ISessionHandler handler, SessionHandlerOptions handlerOptions, ExecutorService executorService) throws InterruptedException, ServiceBusException;
```

Comment:
we will need to update all our samples and document it properly. we can have seperate work item for this.